### PR TITLE
Add `strict-component-boundaries` rule

### PR DESCRIPTION
**Pull request by cartogram**
Another one off this list https://github.com/Shopify/eslint-plugin-shopify/issues/77

### CODE REVIEW

File: docs/rules/strict-component-boundaries.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Prevent module imports between components. (strict-component-boundaries)
+
+Components should have everything they need co-located in a single directory. This makes it easy to move components to a more appropriate spot, and makes it easy for future developers to find, understand, and update everything about a component in one place.
```
Comment (lemonmade) [inline docs/rules/strict-component-boundaries.md]: I think need a comment that other components should not be reaching in/ the component should export its entire API

File: docs/rules/strict-component-boundaries.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+```js
+import {something} from './components/SomeComponent';
+import something from './SomeComponent/anything';
+```
+
+Examples of **correct** code for this rule:
+
+```js
+import {somthing} from './components';
```
Comment (lemonmade) [inline docs/rules/strict-component-boundaries.md]: Missing an e, make it pascalcase

File: tests/lib/rules/strict-component-boundaries.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  valid: [
+    {code: `import {something} from './components';`, parserOptions},
+    {
+      code: `import {something} from 'components';`,
+      parserOptions,
+    },
+  ],
+  invalid: [
+    {
+      code: `import something from '../OtherComponent/any-path';`,
```
Comment (lemonmade) [inline tests/lib/rules/strict-component-boundaries.js]: Can you check that ../OtherComponent works?

File: lib/rules/strict-component-boundaries.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  create(context) {
+    return {
+      ImportDeclaration(node) {
+        const pathParts = node.source.value
+          .split('/')
+          .filter((part) => part[0] !== '.');
+
+        if (
+          (hasAnotherComponentInPath(pathParts) && pathParts.length > 1) ||
+          (hasComponentDirectoryInPath(pathParts) && pathParts.length > 2)
```
Comment (lemonmade) [inline lib/rules/strict-component-boundaries.js:23]: I think this might not catch ‘components/Foo’

**Review by lemonmade — CHANGES_REQUESTED** [review]

File: lib/rules/strict-component-boundaries.js (hunk 4 above)
Comment (cartogram) [reply lib/rules/strict-component-boundaries.js:23]: The first one should catch that https://github.com/Shopify/eslint-plugin-shopify/pull/98/files#diff-f91ed48b12fe78aaf97a3a26453bd862R31

**Comment by cartogram** [comment]
Thanks for the feedback @lemonmade, it is ready for another look now

**Review by lemonmade — APPROVED** [review]

**Pull request merged**